lte/gateway/python/magma/enodebd/devices/baicells_436Q.py

```diff
@@ -0,0 +1,825 @@
+
+    def get_msg(self, message: Any) -> AcsMsgAndTransition:
+        """
+        Send back a message to enb
+
+        Args:
+            message (Any): TR069 message
+
+        Returns:
+            AcsMsgAndTransition
+        """
+        request = models.DummyInput()
+        return AcsMsgAndTransition(msg=request, next_state=self.done_transition)
+
+    def state_description(self) -> str:
+        """
+        Describe the state
+
+        Returns:
+            str
+        """
+        return 'Completed provisioning eNB. Notifying DP.'
+
+
+BAND = 48
+BAND_48_NOFFS_DL = 55240  # noqa
+BAND_48_LOW_FREQ_MHZ = 3550  # noqa
+SAS_BANDWIDTHS = (5, 10, 15, 20)
+MIN_POWER_SPECTRAL_DENSITY = -137
+MAX_POWER_SPECTRAL_DENSITY = 37
+
+
+def update_desired_config_based_on_sas_grant(state: CBSDStateResult, config: EnodebConfiguration) -> None:
+    """
+    Call grpc endpoint on the Domain Proxy to update the desired config based on sas grant
+
+    Args:
+        state (CBSDStateResult): state result as received from DP
+        config (EnodebConfiguration): configuration to update
+    """
+    config.set_parameter(ParameterName.SAS_RADIO_ENABLE, state.radio_enabled)
+    if not state.radio_enabled:
+        return
+
+    earfcn = _calc_earfcn(state.channel.low_frequency_hz, state.channel.high_frequency_hz)
+    bandwidth_mhz = _calc_bandwidth_mhz(state.channel.low_frequency_hz, state.channel.high_frequency_hz)
+    bandwidth_rbs = _calc_bandwidth_rbs(bandwidth_mhz)
+    psd = _calc_psd(state.channel.max_eirp_dbm_mhz)
+
+    config.set_parameter(ParameterName.BAND, BAND)
+    config.set_parameter(ParameterName.DL_BANDWIDTH, bandwidth_rbs)
+    config.set_parameter(ParameterName.UL_BANDWIDTH, bandwidth_rbs)
+    config.set_parameter(ParameterName.EARFCNDL, earfcn)
+    config.set_parameter(ParameterName.EARFCNUL, earfcn)
+    config.set_parameter(ParameterName.POWER_SPECTRAL_DENSITY, psd)
+
+
+def _calc_psd(eirp: float) -> int:
+    psd = int(eirp)
+    if not MIN_POWER_SPECTRAL_DENSITY <= psd <= MAX_POWER_SPECTRAL_DENSITY:  # noqa
+        raise ConfigurationError('Power Spectral Density %d exceeds allowed range [%d, %d]' %
+                                 (psd, MIN_POWER_SPECTRAL_DENSITY, MAX_POWER_SPECTRAL_DENSITY))
+    return psd
+
+
+def _calc_bandwidth_mhz(low_freq_hz: int, high_freq_hz: int) -> float:
+    bandwidth_mhz = (high_freq_hz - low_freq_hz) / 1e6
+    i = bisect_right(SAS_BANDWIDTHS, bandwidth_mhz)
+    if not i:
+        raise ConfigurationError('Unknown/unsupported bandwidth specification (%f)' % bandwidth_mhz)
+    return SAS_BANDWIDTHS[i - 1]
+
+
+def _calc_bandwidth_rbs(bandwidth_mhz: float) -> str:
+    bandwidth_rbs = int(5 * bandwidth_mhz)
+    return str(bandwidth_rbs)
+
+
+def _calc_earfcn(low_freq_hz: int, high_freq_hz: int) -> int:
+    mid_frequency_mhz = (high_freq_hz + low_freq_hz) / 2e6
+    earfcn = 10 * (mid_frequency_mhz - BAND_48_LOW_FREQ_MHZ) + BAND_48_NOFFS_DL
+    return int(earfcn)
```

Comment:
Can we move all of these to a helpers file?

Reply:
Moved the helper functions to utils.
However the model specific method needs to stay in the device model as its only the device model alone know, how to translate Domain Proxy/SAS grant parameters onto actual TR nodes and their values.